# Probability for Machine Learning

Probability helps us measure uncertainty.

Instead of saying an event will definitely happen, probability tells us how likely it is to happen.

Machine Learning models frequently predict probabilities before making decisions.

In this notebook, we'll learn the fundamental probability concepts required for Machine Learning.

# Learning Objectives

By the end of this notebook, you will be able to:

- Define probability.
- Understand experiments, outcomes, sample spaces, and events.
- Calculate simple probabilities.
- Differentiate between independent and dependent events.
- Understand conditional probability.
- Explain Bayes' Theorem intuitively.
- Relate probability to Machine Learning models.

# Why Probability Matters in Machine Learning

Many Machine Learning models work with uncertainty.

Instead of making absolute decisions, they estimate the likelihood of different outcomes.

Examples:

- Will a customer churn?
- Is an email spam?
- Will a loan default?
- Does an image contain a cat?

Understanding probability helps us interpret these predictions correctly.

# What is Probability?

Probability is a numerical measure of how likely an event is to occur.

Its value always lies between:

0 and 1

or equivalently,

0% and 100%.

Where:

- 0 → Impossible event
- 1 → Certain event

In [6]:
print("Impossible Event :", 0)
print("Certain Event    :", 1)
print("50% Probability  :", 0.5)

Impossible Event : 0
Certain Event    : 1
50% Probability  : 0.5


## Examples

| Event | Probability |
|--------|-------------|
| Sun rises tomorrow | Close to 1 |
| Roll a 7 on a six-sided die | 0 |
| Toss a fair coin and get Heads | 0.5 |

Probability represents the chance of an event occurring.

# Experiment

An experiment is any action or process that produces one or more possible outcomes.

Examples:

- Tossing a coin
- Rolling a die
- Drawing a card

Each experiment can produce different results.

In [7]:
experiment = "Roll a Die"

print("Experiment:", experiment)

Experiment: Roll a Die


# Outcome

An outcome is a single possible result of an experiment.

Example:

Experiment:

Roll a die

Possible outcomes are:

1, 2, 3, 4, 5, or 6.

Rolling a 4 is one outcome.

In [8]:
outcomes = [1, 2, 3, 4, 5, 6]

print(outcomes)

[1, 2, 3, 4, 5, 6]


# Sample Space

A Sample Space is the set of all possible outcomes of an experiment.

It is usually represented by the symbol **S**.

Examples:

Rolling a die:

S = {1, 2, 3, 4, 5, 6}

Tossing a coin:

S = {Heads, Tails}

Every probability problem starts by identifying the sample space.

In [9]:
sample_space = [1, 2, 3, 4, 5, 6]

print("Sample Space:", sample_space)

Sample Space: [1, 2, 3, 4, 5, 6]


## Why is the Sample Space Important?

Probability is calculated using the sample space.

If we don't know all possible outcomes, we cannot correctly calculate probabilities.

In Machine Learning, this idea is similar to identifying all possible classes in a classification problem.

# Event

An Event is one or more outcomes from the sample space that we are interested in.

Examples:

Rolling an even number:

{2, 4, 6}

Rolling a number greater than 4:

{5, 6}

An event can contain one outcome or multiple outcomes.

In [10]:
sample_space = [1, 2, 3, 4, 5, 6]

even_numbers = [2, 4, 6]

print("Sample Space :", sample_space)
print("Event        :", even_numbers)

Sample Space : [1, 2, 3, 4, 5, 6]
Event        : [2, 4, 6]


# Probability Formula

Probability is calculated as:

Probability = Favorable Outcomes / Total Possible Outcomes

The probability of any event always lies between:

0 and 1

where:

- 0 means the event is impossible.
- 1 means the event is certain.

In [12]:
favorable_outcomes = 3
total_outcomes = 6

probability = favorable_outcomes / total_outcomes

print("Probability:", probability)

Probability: 0.5


## Example

Experiment:

Roll a fair six-sided die.

Event:

Getting an even number.

Favorable outcomes:

{2, 4, 6}

Total outcomes:

{1, 2, 3, 4, 5, 6}

Probability = 3 / 6 = 0.5

This means there is a 50% chance of rolling an even number.

In [13]:
sample_space = [1, 2, 3, 4, 5, 6]

event = [2, 4, 6]

probability = len(event) / len(sample_space)

print("Probability of getting an even number:", probability)

Probability of getting an even number: 0.5


# Probability in Machine Learning

Probability is used throughout Machine Learning.

Examples:

- Spam Detection

  Probability(Spam) = 0.97

- Loan Approval

  Probability(Default) = 0.18

- Customer Churn

  Probability(Churn) = 0.81

Most classification models estimate probabilities before making the final prediction.

# Independent Events

Two events are independent if the outcome of one event does not affect the outcome of the other.

Examples:

- Tossing a coin
- Rolling a die

The result of the coin toss does not change the result of the die roll.

These events are independent.

In [14]:
coin = ["Heads", "Tails"]

die = [1, 2, 3, 4, 5, 6]

print("Coin Outcomes:", coin)
print("Die Outcomes :", die)

Coin Outcomes: ['Heads', 'Tails']
Die Outcomes : [1, 2, 3, 4, 5, 6]


## Machine Learning Example

Suppose we know:

- A customer's favorite color.
- Today's weather.

Knowing one does not usually help predict the other.

These are examples of independent information.

# Dependent Events

Two events are dependent if one event changes the probability of the other.

Example:

Drawing two cards from a deck without replacement.

After the first card is removed, the deck changes.

Therefore, the probability of the second draw also changes.

## Machine Learning Example

Suppose we know:

- A customer recently missed three loan payments.

This information changes the probability that the customer may default in the future.

Here, one piece of information influences another, making the events dependent.

# Conditional Probability

Conditional Probability is the probability of an event occurring given that another event has already occurred.

The additional information changes our prediction.

Example:

Instead of asking,

"What is the probability that a student passes?"

we ask,

"What is the probability that the student passes **given** they attended every lecture?"

The extra information changes the probability.

## Why Conditional Probability Matters in Machine Learning

Machine Learning models rarely make predictions without context.

Instead, they use available information such as:

- Age
- Income
- Education
- Previous purchases
- Credit history

to estimate the probability of an outcome.

The prediction depends on the information provided.

# Bayes' Theorem (Intuition)

Bayes' Theorem helps us update a probability when new evidence becomes available.

Initial belief:

A patient may have a disease.

New evidence:

The medical test result is positive.

Using this new evidence, we update the probability that the patient actually has the disease.

Machine Learning applies this same idea in the Naive Bayes algorithm.

In [15]:
print("Initial Probability of Disease : 0.10")
print("Positive Test Result Received")
print("Updated Probability Changes")

Initial Probability of Disease : 0.10
Positive Test Result Received
Updated Probability Changes


## Where Will We Use This Later?

We will revisit these ideas when learning:

- Naive Bayes
- Logistic Regression
- Classification Models
- Model Probabilities (`predict_proba()`)

For now, it is enough to understand the intuition.

The mathematical details will be introduced only when they become necessary.

# Probability in Scikit-learn

Many classification models can estimate probabilities.

For example,

`predict_proba()`

returns the probability of each class.

The class with the highest probability becomes the predicted class.

In [16]:
import pandas as pd

prediction = pd.DataFrame(
    {
        "Not Spam": [0.08],
        "Spam": [0.92]
    }
)

prediction

,Not Spam,Spam
0,0.08,0.92


The model predicts:

- 8% probability that the email is **Not Spam**
- 92% probability that the email is **Spam**

Since 92% is higher, the model predicts **Spam**.

# Key Takeaways

- Probability measures how likely an event is to occur.
- Probability values always lie between 0 and 1.
- Every probability problem begins with a sample space.
- An event is one or more outcomes from the sample space.
- Independent events do not affect each other.
- Dependent events influence each other.
- Conditional probability uses additional information to update predictions.
- Many Machine Learning models estimate probabilities before making predictions.

# Notebook Summary

In this notebook, we learned:

- Probability
- Experiment
- Outcome
- Sample Space
- Event
- Probability Formula
- Independent Events
- Dependent Events
- Conditional Probability
- Bayes' Theorem (Intuition)
- Probability in Machine Learning

These concepts provide the probability foundation required for classification algorithms such as Naive Bayes and Logistic Regression, as well as for interpreting prediction probabilities from Machine Learning models.